# 4.4 - Practice work 

# Importing libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# Importing Data

In [2]:
path = r'C:\Users\srira\OneDrive\Documents\Instacart Basket Analysis'

In [4]:
# Importing data file - products.csv
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Importing data file - orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Data Consistency checks

In [6]:
# Returns descriptive statistics and also checks for the accuracy of the columns in the dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_days_of_week  order_hour_of_day  days_since_prior_order  
count         3.421083e+06       3.421083e+06            3.214874e+06  
mean          2.776219e+00       1.345202e+01            1.111484e+01  
std           2.046829e+00       4.226088e+00            9.206737e+00  
min           0.000000e+00       0.000000e+00            0.000000e+00  
25%           1.000000e+00       1.000000e+01            4.000000e+00  
50%           3.000000e+00       1.300000e+01            7.000000e+00  
75%           5.000000e+00       1.600000e+01            1.500000e+01  
max           6.000000e+00       2.300000e+01            3.000000e+01

# Mixed - type data

In [7]:
# Create a new test dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', '1', 'true']

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  true

In [47]:
# Check for mixed data types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)). any(axis = 1)
    if len(df_test[weird]) > 0:
        print(col)

# Missing values

In [16]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
# Displays the records where the product_name is Null
df_nan = df_prods[df_prods['product_name']. isnull() == True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [ ]:
# Addressing Missing values

In [20]:
# Checking the number of rows in df_prods dataframe
df_prods.shape

(49693, 5)

In [24]:
# Creating a new dataframe for non missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
# Current number of rows in the new dataframe after addressing the missing values
df_prods_clean.shape

(49677, 5)

# Duplicates

In [26]:
# Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [27]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [ ]:
# Addressing duplicate values

In [28]:
# Current number of rows in the dataframe
df_prods_clean.shape

(49677, 5)

In [29]:
# Removed the duplicates and saved in a new dataframe 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [31]:
# Current number of rows in the new dataframe after removing the duplicates
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5 - Exercise

# Question 2: 

In [32]:
# Looking the descriptive statistics and accuracy of the dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_days_of_week  order_hour_of_day  days_since_prior_order  
count         3.421083e+06       3.421083e+06            3.214874e+06  
mean          2.776219e+00       1.345202e+01            1.111484e+01  
std           2.046829e+00       4.226088e+00            9.206737e+00  
min           0.000000e+00       0.000000e+00            0.000000e+00  
25%           1.000000e+00       1.000000e+01            4.000000e+00  
50%           3.000000e+00       1.300000e+01            7.000000e+00  
75%           5.000000e+00       1.600000e+01            1.500000e+01  
max           6.000000e+00       2.300000e+01            3.000000e+01

Answer: When looking into the above values, especially in the min and max, firstly in the order_number column, the min and max values are 1 and 1 but the values of the percentile(25% is 5 and 75% is 2.3) doesn't seem to be consistent. Similarly in the days_since_prior_order column, the min and max values are 0 and 3 but the values of the percentiles are  4 and 7 which is inconsistent.

# Question 3: Checking for mixed data types

In [34]:
# Check for mixed data type in the df_ords dataframe
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_days_of_week  \
0           0   2539329        1             1                    2   
1           1   2398795        1             2                    3   
2           2    473747        1             3                    3   
3           3   2254736        1             4                    4   
4           4    431534        1             5                    4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [35]:
# Finding mixed data types
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)). any(axis = 1)
    if len(df_ords[weird]) > 0:
        print(col)

 Answer: There are no mixed data types in the df_ords dataframe.

# Question 4: Checking for columns with mixed data types

Answer : There are no columns with mixed data types in the df_ords dataframe.

# Question 5: Finding Missing values

In [36]:
# Checking for missing values in the df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_days_of_week            0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Answer: There are 206209 missing values in the column "days_since_prior_order" in the df_ords dataframe.

# Question 6: Addressing missing values

In [37]:
# Addressing the missing values in the df_ords dataframe
df_ords_nan = df_ords[df_ords['days_since_prior_order']. isnull() == True]

In [38]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_days_of_week  \
0                 0   2539329        1             1                    2   
11               11   2168274        2             1                    2   
26               26   1374495        3             1                    1   
39               39   3343014        4             1                    6   
45               45   2717275        5             1                    3   
...             ...       ...      ...           ...                  ...   
3420930     3420930    969311   206205             1                    4   
3420934     3420934   3189322   206206             1                    3   
3421002     3421002   2166133   206207             1                    6   
3421019     3421019   2227043   206208             1                    1   
3421069     3421069   3154581   206209             1                    3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Answer: In the df_ords dataframe, there are considerably a large number of missing values. This is because when a customer places the first order, then the days_since_prior_order would not be applicable as the customer is making the first order. All the null values are displayed for users when their order_number is 1. Hence these values are relevant and need not be deleted or imputed using any of the methods, as the days_since_prior_order column will automatically get updated once when the customer makes more orders.

# Question 7: Checking for duplicates

In [39]:
# Finding duplicate values in the df_ords dataframe
df_ords_dups = df_ords[df_ords.duplicated()]

In [40]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_days_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Answer: No duplicates were found in the df_ords dataframe.

# Question 8: Addressing duplicates

In [ ]:
Answer: As no duplicates were found in the df_ords dataframe, there would be no need to address them.

# Question 9: Exporting data

In [43]:
# Exporting df_ords as orders_checked.csv
df_ords.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_checked.csv'))

In [45]:
# Exporting df_prods as products_checked.csv
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'products_checked.csv'))